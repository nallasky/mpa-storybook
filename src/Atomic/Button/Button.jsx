import React from "react";
import { css } from "@emotion/react";
import PropTypes from "prop-types";

/**
 * Button 컴포넌트는 어떠한 작업을 트리거 할 때 사용합니다.
 */
const Button = ({ children, onClick, theme, size, disabled, width, iconOnly, style }) => {
	return (
		<button
			css={[
				defStyle,
				themes[theme],
				sizes[size],
				{ width },
				iconOnly && [iconOnlyStyle, iconOnlySizes[size]],
				style
			]}
			onClick={onClick}
			disabled={disabled}
		>
			{children}
		</button>
	);
};

Button.propTypes = {
  /**
   * 버튼 안에 표시할 내용을 설정합니다.
   */
  children: PropTypes.node.isRequired,
  /**
   * 버튼 클릭 시 호출되는 함수를 설정합니다.
   */
  onClick: PropTypes.func,
  /**
   * 버튼의 테마를 설정합니다.
   */
  theme: PropTypes.oneOf(['blue', 'green', 'orange', 'white']),
  /**
   * 버튼의 크기를 설정합니다.
   */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /**
   * 버튼의 활성화 여부를 설정합니다.
   */
  disabled: PropTypes.bool,
  /**
   * 버튼의 너비를 설정합니다.
   */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * 버튼 안에 아이콘만 표시할 때 설정합니다.
   */
  iconOnly: PropTypes.bool,
  /**
   * 버튼에 커스텀 스타일을 적용할 때 설정합니다.
   */
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
};

Button.defaultProps = {
	theme: 'blue',
	size: 'sm'
};

const defStyle = css`
	outline: none;
	border: none;
	color: white;
	box-sizing: border-box;
	border-radius: .5rem;
	height: 2rem;
	font-family: Saira, "Nunito Sans";
	font-size: 0.875rem;
	padding: 0 1rem;
	font-weight: 600;
	display: inline-flex;
	align-items: center;
	justify-content: center;
	box-shadow: 0px 3px 3px rgba(0,0,0,0.25);
	&:active:enabled {
    transform: translateY(2px);
    box-shadow: none;
	}
  &:hover:not(:disabled) {
    cursor: pointer;
  }
	&:disabled {
    cursor: not-allowed;
	}
  svg {
    fill: currentColor;
    width: 1em;
    margin-right: 1em;
  }
`;

const themes = {
  blue: css`
    background: rgb(24, 144, 255);
    &:hover:not(:disabled) {
      background: rgba(24, 144, 255, 0.8);
    }
    &:disabled {
      background: rgba(24, 144, 255, 0.4);
    }
  `,
  green: css`
    background: rgb(60, 191, 100);
    &:hover:not(:disabled) {
      background: rgba(60, 191, 100, 0.8);
    }
    &:disabled {
      background: rgba(60, 191, 100, 0.4);
    }
  `,
  orange: css`
    background: rgb(232, 92, 25);
    &:hover:not(:disabled) {
      background: rgba(232, 92, 25, 0.8);
    }
    &:disabled {
      background: rgba(232, 92, 25, 0.4);
    }
  `,
  white: css`
    background: white;
    color: rgba(0, 0, 0, 0.85);
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0px 3px 3px rgba(0, 0, 0, 0.1);
    &:hover:not(:disabled) {
      background: rgba(0, 0, 0, 0.05);
    }
    &:disabled {
      color: rgba(0, 0, 0, 0.2);
      background: rgba(0, 0, 0, 0.04);
    }
  `
};

const sizes = {
  sm: css`
	  height: 1.75rem;
	  font-size: 0.75rem;
	  padding: 0 0.875rem;
  `,
  md: css`
	  height: 2.5rem;
	  font-size: 1rem;
	  padding: 0 1rem;
  `,
  lg: css`
	  height: 3rem;
	  font-size: 1.125rem;
	  padding: 0 1.5rem;
  `
};

const iconOnlyStyle = css`
  padding: 0;
  border-radius: 50%;
  svg {
    margin: 0;
  }
`;

const iconOnlySizes = {
  sm: css`
    width: 1.75rem;
  `,
  md: css`
    width: 2.5rem;
  `,
  lg: css`
    width: 3rem;
  `
};

export default Button;