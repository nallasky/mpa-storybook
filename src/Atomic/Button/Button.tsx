/** @jsx jsx */
import { jsx, css } from '@emotion/core';

type ButtonProps = {
	/** 버튼 안의 내용 */
	children: React.ReactNode;
	/** 클릭 시 호출되는 함수 */
	onClick?: (e?: React.MouseEvent<HTMLButtonElement>) => void;
	/** 버튼의 테마를 설정합니다. */
	theme: 'blue' | 'green' | 'orange';
	/** 버튼의 크기를 설정합니다. */
	size: 'sm' | 'md' | 'lg';
	/** 버튼을 비활성화 시킵니다. */
	disabled?: boolean;
	/** 버튼의 너비를 설정합니다. */
	width?: string | number;
};

/** Button 컴포넌트는 어떠한 작업을 트리거 할 때 사용합니다. */
const Button = ({ children, onClick, theme, size, disabled, width }: ButtonProps) => {
	return (
		<button
				css={[style, themes[theme], sizes[size], { width }]}
				onClick={onClick}
				disabled={disabled}
		>
			{children}
		</button>
	);
};

Button.defaultProps = {
	theme: 'blue',
	size: 'sm'
};

const style = css`
	outline: none;
	border: none;
	color: white;
	box-sizing: border-box;
	border-radius: .25rem;
  height: 2rem;
  font-size: 0.875rem;
  padding: 0 1rem;
  font-weight: 600;
  display: inline-flex;
  align-items: center;
  justify-content: center;
	transition: transform .15s, box-shadow .15s;
	box-shadow: 0px 3px 3px rgba(0,0,0,0.25);
	&:active:enabled {
		transform: translateY(2px);
		box-shadow: none;
	}
	&:disabled {
		cursor: not-allowed;
	}
`;

const themes = {
	blue: css`
    background: rgb(24, 144, 255);
    &:hover:enabled {
      background: rgba(24, 144, 255, .8);
    }
		&:disabled {
			background: rgba(24, 144, 255, .4);
		}
  `,
	green: css`
    background: rgb(60, 191, 100);
    &:hover:enabled {
      background: rgba(60, 191, 100, .8);
    }
		&:disabled {
			background: rgba(60, 191, 100, .4);
		}
	`,
	orange: css`
		background: rgb(232, 92, 25);
    &:hover:enabled {
      background: rgba(232, 92, 25, .8);
    }
		&:disabled {
			background: rgba(232, 92, 25, .4);
		}
	`
};

const sizes = {
	sm: css`
    height: 1.75rem;
    font-size: 0.75rem;
    padding: 0 0.875rem;
  `,
	md: css`
    height: 2.5rem;
    font-size: 1rem;
    padding: 0 1rem;
  `,
	lg: css`
    height: 3rem;
    font-size: 1.125rem;
    padding: 0 1.5rem;
  `
};

export default Button;