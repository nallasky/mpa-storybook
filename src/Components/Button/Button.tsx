/** @jsx jsx */
import { jsx, css } from '@emotion/core';

type ButtonProps = {
	/** 버튼 안의 내용 */
	children: React.ReactNode;
	/** 클릭 시 호출되는 함수 */
	onClick?: (e?: React.MouseEvent<HTMLButtonElement>) => void;
};

/** Button 컴포넌트는 어떠한 작업을 트리거 할 때 사용합니다. */
const Button = ({ children, onClick }: ButtonProps) => {
	return (
		<button css={style} onClick={onClick}>
			{children}
		</button>
	);
};

const style = css`
	outline: none;
	border: none;
	box-sizing: border-box;
	font-size: .875rem;
	padding: .5rem 1rem;
	background: #20c997;
	color: white;
	border-radius: .25rem;
	line-height: 1;
	font-weight: 600;
	&:focus {
		box-shadow: 0 0 8px rgba(0,0,0,0.2);
	}
	&:hover {
		background: #38d9a9;
	}
	&:active {
		background: #12b886;
	}
`;

export default Button;