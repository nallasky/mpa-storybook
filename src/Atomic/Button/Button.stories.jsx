import React from "react";
import Button from './Button';
import { css } from '@emotion/react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAddressCard, faBandAid, faBell, faBolt, faClock } from "@fortawesome/free-solid-svg-icons";
import { withTests } from "@storybook/addon-jest";
import results from '../../.jest-test-results.json';

export default {
  title: 'Atomic/Button',
  component: Button,
  decorators: [withTests({ results })]
};

const Template = (args) => {
	return (
		<Button
			theme={args.theme}
			size={args.size}
			disabled={args.disabled}
			width={args.width}
			iconOnly={args.iconOnly}
			style={args.style}
			onClick={args.onClick}
		>
			{args.children}
		</Button>
	);
};

export const Basic = Template.bind({});
Basic.args = {
	children: "BUTTON",
	theme: "blue",
	size: "sm"
};

const buttonWrapper = css`
  .description {
    margin-bottom: 0.5rem;
    font-family: Saira;
    font-weight: 600;
  }
  & > div + div {
    margin-top: 2rem;
  }
`;

export const Themes = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <div className="description">Default(Blue)</div>
        <Button>BUTTON</Button>
      </div>
      <div>
        <div className="description">Green</div>
        <Button theme="green">BUTTON</Button>
      </div>
      <div>
        <div className="description">Orange</div>
        <Button theme="orange">BUTTON</Button>
      </div>
      <div>
          <div className="description">White</div>
        <Button theme="white">BUTTON</Button>
      </div>
    </div>
  );
};

export const Sizes = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <div className="description">Small</div>
        <Button size="sm">BUTTON</Button>
      </div>
      <div>
        <div className="description">Medium</div>
        <Button size="md">BUTTON</Button>
      </div>
      <div>
        <div className="description">Big</div>
        <Button size="lg">BUTTON</Button>
      </div>
    </div>
  );
};

export const Disabled = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <Button disabled>BUTTON</Button>
      </div>
      <div>
        <Button disabled theme="green">
          BUTTON
        </Button>
      </div>
      <div>
        <Button disabled theme="orange">
          BUTTON
        </Button>
      </div>
      <div>
        <Button disabled theme="white">
          BUTTON
        </Button>
      </div>
    </div>
  );
};

export const Width = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <Button width="20rem">CUSTOM WIDTH</Button>
      </div>
      <div>
        <Button width="100%">FULL WIDTH</Button>
      </div>
    </div>
  );
};

export const WithIcon = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <Button>
          <FontAwesomeIcon icon={faAddressCard} /> BUTTON
        </Button>
      </div>
      <div>
        <Button theme="green">
          <FontAwesomeIcon icon={faBandAid} /> BUTTON
        </Button>
      </div>
      <div>
        <Button theme="orange">
          <FontAwesomeIcon icon={faBell} /> BUTTON
        </Button>
      </div>
      <div>
        <Button theme="white">
          <FontAwesomeIcon icon={faBolt} /> BUTTON
        </Button>
      </div>
    </div>
  );
};

export const IconOnly = () => {
  return (
    <div css={buttonWrapper}>
      <div>
        <Button iconOnly>
          <FontAwesomeIcon icon={faClock} />
        </Button>
      </div>
      <div>
        <Button iconOnly size="md" theme="green">
          <FontAwesomeIcon icon={faClock} />
        </Button>
      </div>
      <div>
        <Button iconOnly size="lg" theme="orange">
          <FontAwesomeIcon icon={faClock} />
        </Button>
      </div>
      <div>
        <Button iconOnly size="lg" theme="white">
          <FontAwesomeIcon icon={faClock} />
        </Button>
      </div>
    </div>
  );
};