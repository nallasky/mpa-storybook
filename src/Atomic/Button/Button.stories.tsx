/** @jsx jsx */
import React from 'react';
import Button from './Button';
import { withKnobs, text, boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { jsx, css } from '@emotion/core';
import Icon from '../../Icon/Icon';

export default {
	title: 'Atomic|Button',
	component: Button,
	decorators: [withKnobs]
};

export const button = () => {
	const label = text('children', 'BUTTON');
	const size = select('size', ['sm', 'md', 'lg'], 'sm');
	const theme = select(
			'theme',
			['blue', 'green', 'orange', 'white'],
			'blue'
	);
	const disabled = boolean('disabled', false);
	const width = text('width', '');

	return (
			<Button
					size={size}
					theme={theme}
					disabled={disabled}
					width={width}
					onClick={action('onClick')}
			>
				{label}
			</Button>
	);
};

button.story = {
	name: 'Default'
};

export const blueButton = () => {
	return <Button theme="blue">BUTTON</Button>;
};

export const greenButton = () => {
	return <Button theme="green">BUTTON</Button>;
};

export const orangeButton = () => {
	return <Button theme="orange">BUTTON</Button>;
};

export const whiteButton = () => {
	return <Button theme="white">BUTTON</Button>;
};

const buttonWrapper = css`
  .description {
    margin-bottom: 0.5rem;
    font-family: "Nunito Sans",-apple-system,".SFNSText-Regular","San Francisco",BlinkMacSystemFont,"Segoe UI","Helvetica Neue",Helvetica,Arial,sans-serif;
    font-weight: 600;
  }
  & > div + div {
    margin-top: 2rem;
  }
`;

export const sizes = () => {
	return (
			<div css={buttonWrapper}>
				<div>
					<div className="description">Small</div>
					<Button size="sm">BUTTON</Button>
				</div>
				<div>
					<div className="description">Medium</div>
					<Button size="md">BUTTON</Button>
				</div>
				<div>
					<div className="description">Big</div>
					<Button size="lg">BUTTON</Button>
				</div>
			</div>
	);
};

export const disabled = () => {
	return (
			<div css={buttonWrapper}>
				<div>
					<Button disabled>BUTTON</Button>
				</div>
				<div>
					<Button disabled theme="green">
						BUTTON
					</Button>
				</div>
				<div>
					<Button disabled theme="orange">
						BUTTON
					</Button>
				</div>
				<div>
					<Button disabled theme="white">
						BUTTON
					</Button>
				</div>
			</div>
	);
};

export const customSized = () => {
	return (
			<div css={buttonWrapper}>
				<div>
					<Button width="20rem">CUSTOM WIDTH</Button>
				</div>
				<div>
					<Button width="100%">FULL WIDTH</Button>
				</div>
			</div>
	);
};

export const withIcon = () => {
	return (
			<div css={buttonWrapper}>
				<div>
					<Button>
						<Icon icon="setting" /> BUTTON
					</Button>
				</div>
				<div>
					<Button theme="green">
						<Icon icon="setting" /> BUTTON
					</Button>
				</div>
				<div>
					<Button theme="orange">
						<Icon icon="heart" /> BUTTON
					</Button>
				</div>
				<div>
					<Button theme="white">
						<Icon icon="binoculars" /> BUTTON
					</Button>
				</div>
			</div>
	);
};

export const iconOnly = () => {
	return (
			<div css={buttonWrapper}>
				<div>
					<Button iconOnly>
						<Icon icon="heart" />
					</Button>
				</div>
				<div>
					<Button iconOnly size="md" theme="green">
						<Icon icon="heart" />
					</Button>
				</div>
				<div>
					<Button iconOnly size="lg" theme="orange">
						<Icon icon="heart" />
					</Button>
				</div>
			</div>
	);
};