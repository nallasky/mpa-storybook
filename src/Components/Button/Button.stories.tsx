import React from 'react';
import Button from './Button';
import { withKnobs, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

export default {
	title: 'components|Button',
	component: Button,
	decorators: [withKnobs]
};

export const button = () => {
	const children = text('children', 'BUTTON');
	return <Button onClick={action('onClick')}>{children}</Button>;
};

button.story = {
	name: 'Default'
};

export const primaryButton = () => {
	return <Button>PRIMARY</Button>;
};