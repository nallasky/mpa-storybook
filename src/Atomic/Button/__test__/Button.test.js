import React from 'react';
import Button from '../Button';
import { mount } from "enzyme";

describe('Button test', () => {
  it('matches snapshot (initialProp)', () => {
    const wrapper = mount(<Button>Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (green theme)', () => {
    const wrapper = mount(<Button theme="green">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (orange theme)', () => {
    const wrapper = mount(<Button theme="orange">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (white theme)', () => {
    const wrapper = mount(<Button theme="white">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (md size)', () => {
    const wrapper = mount(<Button size="md">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (lg size)', () => {
    const wrapper = mount(<Button size="lg">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (disabled)', () => {
    const wrapper = mount(<Button disabled>Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (change width)', () => {
    const wrapper = mount(<Button width="30rem">Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (icon only)', () => {
    const wrapper = mount(<Button iconOnly>Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('check onClick function', () => {
    const clickFunc = jest.fn();
    const wrapper = mount(<Button onClick={clickFunc}>Test</Button>);
    expect(clickFunc).not.toHaveBeenCalled();

    wrapper.find('button').simulate('click');
    expect(clickFunc).toHaveBeenCalledTimes(1);
  });
});