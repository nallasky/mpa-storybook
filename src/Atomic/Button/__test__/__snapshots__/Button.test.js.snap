// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test matches snapshot (change width) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (disabled) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (green theme) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (icon only) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (initialProp) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (lg size) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (md size) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (orange theme) 1`] = `ReactWrapper {}`;

exports[`Button test matches snapshot (white theme) 1`] = `ReactWrapper {}`;
